polaris-react/src/components/IndexTable/IndexTable.tsx

```diff
@@ -910,11 +915,12 @@ function IndexTableBase({
         !isCurrentlySorted && index === lastSortedColumnIndex.current;
 
       const isAscending = sortDirection === 'ascending';
-      let newDirection: IndexTableSortDirection = defaultSortDirection;
+
+      let newDirection: IndexTableSortDirection =
+        heading.defaultSortDirection || defaultSortDirection;
```

Comment:
```suggestion
        heading.defaultSortDirection ?? defaultSortDirection;
```

(my initial suggestion in Slack was wrong, we need ?? here)